course/models.py

```diff
@@ -1704,15 +1719,55 @@ def consume(self, iterable, set_is_superseded=False):
             gchange.is_superseded = False
             self._consume_grade_change(gchange, set_is_superseded)
 
-        valid_grade_changes = sorted(
-                (gchange
-                for gchange in self.attempt_id_to_gchange.values()
-                if gchange.percentage() is not None),
-                key=lambda gchange: gchange.grade_time)
+        gchanges = [gchange for gchange in self.attempt_id_to_gchange.values()]
+
+        valid_grade_changes = sorted(gchanges, key=lambda x: (x.grade_time, x.pk))
+
+        import copy
+        _valid_percentages_copy = copy.copy(self.valid_percentages)
 
         self.valid_percentages.extend(
-                cast(GradeChange, gchange.percentage())
-                for gchange in valid_grade_changes)
+                [gchange.percentage() for gchange in valid_grade_changes])
+
+        # {{{ Calculate the earliest percentage and latest percentage, instead of
+        # get them from self.valid_percentages, so as to avoid inconsistent results.
+
+        if self.opportunity is not None:
+            strategy = self.opportunity.aggregation_strategy
+            if self.valid_percentages and strategy in [
+                            grade_aggregation_strategy.use_earliest,
+                            grade_aggregation_strategy.use_latest]:
+
+                gchanges_with_session = sorted([
+                    gchange for gchange in gchanges
+                    if gchange.flow_session is not None],
+                    key=lambda x: (x.effective_time, x.pk))
+
+                # There are no more than 1 flow session, just return the grade
+                # changes ordered by grade_time.
+                if len(gchanges_with_session) <= 1:
+                    self.earliest_percentage = self.valid_percentages[0]
+                    self.latest_percentage = self.valid_percentages[-1]
+                else:
+                    idx = (
+                        0 if strategy == grade_aggregation_strategy.use_earliest
+                        else -1)
+
+                    _gchanges = sorted(
+                        [gchange for gchange in gchanges
+                         if gchange.flow_session is None]
+                        + [gchanges_with_session[idx]],
+                        key=lambda x: (x.grade_time, x.pk))
+
+                    _valid_percentages = (
+                        _valid_percentages_copy +
+                        [gchange.percentage() for gchange in _gchanges])
+
+                    if strategy == grade_aggregation_strategy.use_earliest:
+                        self.earliest_percentage = _valid_percentages[0]
+                    else:
+                        self.latest_percentage = _valid_percentages[-1]
```

Comment:
Why is all this still needed?

Reply:
I must admit the logic is not clear enough.

We still have the trouble after introducing `effective_time`:
Now only session-related grade_changes were using that attribute (For non-session-related ones, it's null). This is done intended. For comparing two gchange objects:
1. If either one is not session-related, use `grade_time` as key, then pk.
1. If both are session-related, use `effective_time`, then pk.

We can't have the above situation both use `effective_time`, because once a grade has been imported, the regrades previous submitted session will be overriden.

I'll try to simply the logic above.

Reply:
Inconsistency were observed for a610287d30a85e79be051bbc757099579cd260e6, with order using using `__eq__ `, `__ge__`, `__le__` in `GradeChange` model.

Reply:
See a610287d30a85e79be051bbc757099579cd260e6, tests [passed](https://ci.appveyor.com/project/inducer/relate/build/1.0.517) and [failed](https://ci.appveyor.com/project/dzhuang/relate/build/1.0.857). The [failed test](https://github.com/inducer/relate/blob/a610287d30a85e79be051bbc757099579cd260e6/tests/test_grades/test_grade_book.py#L469-L503) is actually the one you suggested in #423:

>> Can you give an example?

> I'll try:

> A: grade time 2015
> B: grade time 2016, flow completed 2014
> C: grade time 2017
> D: grade time 2018, flow completed 2013
> Then A < B < C

> But also D < B and C < D, which is a direct contradiction.

Reply:
Updated. Removed `self.earliest_percentage` and `self.latest_percentage`